webpack/components/RoutedTabs/RoutedTabs.js

```diff
@@ -0,0 +1,89 @@
+import React, { useEffect } from 'react';
+import { shape, string, number, element, arrayOf } from 'prop-types';
+import { Tab, Tabs, TabTitleText } from '@patternfly/react-core';
+import { useHistory, useLocation } from 'react-router-dom';
+import './RoutedTabs.scss';
+import paramsFromHash from '../../utils/paramsFromHash';
+
+/*
+  Creates tabs that change url hash based on the tab selected.
+  Also can show a subpage within a tab based on query params.
+  The url format is: "/my/base/url#currentTab?subContentId=1"
+  Influenced by https://github.com/ansible/awx/blob/devel/awx/ui_next/src/components/RoutedTabs/RoutedTabs.jsx
+*/
+const RoutedTabs = ({ tabs, baseUrl, defaultTabIndex }) => {
+  const history = useHistory();
+  const { hash } = useLocation();
+  const { hash: tabFromUrl, params: { subContentId } } = paramsFromHash(hash);
+
+  const buildLink = tabKey => `${baseUrl}#${tabKey}`;
+
+  const changeTab = (eventKey) => {
+    const matchedTab = tabs.find(tab => tab.key === eventKey);
+    if (matchedTab) {
+      history.push(buildLink(matchedTab.key));
+    } else {
+      history.push(buildLink(tabs[defaultTabIndex].key)); // go to first tab if no tab selected
+    }
+  };
+
+  const handleTabSelect = (_event, eventKey) => changeTab(eventKey);
+
+  const getActiveTab = () => {
+    const matchedTab = tabs.find(tab => tab.key === tabFromUrl);
+    if (matchedTab) return matchedTab.key;
+
+    return tabs[defaultTabIndex].key; // Default to first tab
+  };
+
+  // Useful when first navigating to the page, switches to default tab in url
+  useEffect(() => {
+    if (tabFromUrl !== getActiveTab()) changeTab(tabFromUrl);
+  }, [tabFromUrl]);
+
+  // Handle subroutes to show item's detail content while staying on a tab
+  const showContent = (tab) => {
+    const { content, detailContent } = tab;
+    if (subContentId && detailContent) return detailContent;
+
+    return content; // show main content if no subroute
+  };
+
+  return (
+    <Tabs activeKey={getActiveTab()} onSelect={handleTabSelect}>
+      {tabs.map((tab) => {
+        const { key, title } = tab;
+
+        return (
+          <Tab
+            aria-label={title}
+            eventKey={key}
+            key={key}
+            link={`${baseUrl}#${key}`}
```

Comment:
If you pass `href` instead of `link` here, the tab will render as an `<a>` rather than a `<button>`.  Thoughts on changing this?

The differences are
* As it is now, I get `@@router/LOCATION_CHANGE` Redux actions when I click Back and Forward, but not when I click one of the tabs.
* If it's rendered as an `<a>`, I get `@@router/LOCATION_CHANGE` actions both on click and on browser navigation.
* `<a>` feels slightly more correct since clicking it changes the URL in the URL bar and navigates you to another place.  But I can also see an argument for the other side.
* The CSS styling would have to be overridden (rendering as `<a>` makes every tab underlined.)

More food for thought: https://css-tricks.com/use-button-element/#when-links-make-more-sense

Reply:
I think its a good idea but I just keep running into issues with patternfly that way, here is what happens when I just add a `className` to the tab, without any change to the stylesheet:

![hi](http://www.clipular.com/c/6748262549422080.png?k=-tutWYuD_lCWEok55wHtB2Zi_C4)

full change (remove the className prop makes it go back to normal with underlined links)
```diff
diff --git a/webpack/components/RoutedTabs/RoutedTabs.js b/webpack/components/RoutedTabs/RoutedTabs.js
index 9f73c52..93f66b7 100644
--- a/webpack/components/RoutedTabs/RoutedTabs.js
+++ b/webpack/components/RoutedTabs/RoutedTabs.js
@@ -1,9 +1,9 @@
 import React, { useEffect } from 'react';
 import { shape, string, number, element, arrayOf } from 'prop-types';
-import { Tab, Tabs, TabTitleText } from '@patternfly/react-core';
+import { Tab, Tabs, TabsComponent, TabTitleText } from '@patternfly/react-core';
 import { useHistory, useLocation } from 'react-router-dom';
-import './RoutedTabs.scss';
 import paramsFromHash from '../../utils/paramsFromHash';
+import './RoutedTabs.scss';
 
 /*
   Creates tabs that change url hash based on the tab selected.
@@ -50,16 +50,17 @@ const RoutedTabs = ({ tabs, baseUrl, defaultTabIndex }) => {
   };
 
   return (
-    <Tabs activeKey={getActiveTab()} onSelect={handleTabSelect}>
+    <Tabs activeKey={getActiveTab()} onSelect={handleTabSelect} component={TabsComponent.nav}>
       {tabs.map((tab) => {
         const { key, title } = tab;
 
         return (
           <Tab
             aria-label={title}
+            className="kplain-tab-anchors"
             eventKey={key}
             key={key}
-            link={`${baseUrl}#${key}`}
+            href={`${baseUrl}#${key}`}
             title={<TabTitleText>{title}</TabTitleText>}
           >
             <div className="tab-body-with-spacing">
```

Reply:
In order for `className` to work it would have to be passed down to a native HTML element.  Looks like Patternfly doesn't pass down the `className` in this way for `<Tab>`.

Reply:
also I'd be curious to see what the console error is when you get the blank screen :)

Reply:
In addition to the edited comment above, another option would be just to change the SCSS file for this component to find the tabs without relying on class names, and style as we wish.

Reply:
all of that said, don't spend too much time on this.  I'm fine with keeping as is if it's going to take too long.

Reply:
The pf4 docs say `className` should `additional classes added to the Tab`, but it feels like it's overriding their own. I think passing in any of the pf4 class names would be brittle, they could change outside of our control and we would have to keep in-sync

> all of that said, don't spend too much time on this. I'm fine with keeping as is if it's going to take too long.

I'm leaning this way, I tried the href's when I started and gave up then too :man_shrugging: